[HPR-574] Add documentation for configuring Packer for HCP Packer (#12022)
hashicorp/packer: website/content/docs/configure-hcp-packer.mdx

You can use HCP Packer with both JSON and HCL2 templates. If you are using JSON templates we recommend getting started with 
the [HCP Packer environment variables](#hcp-packer-environment-variables) and then migrating to HCL when possible. 

This page summarizes the methods you can use to connect JSON and HCL2 templates to the HCP Packer registry. It also provides a full list of HCP Packer environment variables. Refer to the [HCP Packer Template Configuration](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration) page in the HCP Packer documentation for full configuration details and examples.

## HCP Packer Environment Variables

The following environment variables let you configure Packer to push image metadata to an active registry without changing your template. You can use environment variables with both JSON and HCL2 templates. Refer to [Basic Configuration With Environment Variables](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration#basic-configuration-with-environment-variables) in the HCP Packer documentation for complete instructions and examples. 

> You must set following environment variables to enable Packer to push metadata to a registry.

Author: ```suggestion
You must set the following environment variables to enable Packer to push metadata to a registry.
```

---
description: |

>   You can configure Packer to publish metadata for all completed builds to an active HCP Packer registry. 
>   Legacy JSON templates can connect to the registry using environment variables. HCL2 templates can connect using an hcp_packer_registry block. 

Reviewer: ```suggestion
   Packer can publish metadata for completed builds to an HCP Packer registry. Legacy JSON templates can connect to the registry using environment variables. HCL2 templates can connect using an hcp_packer_registry block. 
```

page_title: Configuring HCP Packer
---

# Configuring HCP Packer

[HCP Packer](https://cloud.hashicorp.com/docs/packer) stores metadata about your Packer images so that you can track updates, 
use the most up-to-date base images, and deploy the most up-to-date downstream images. 

> HCP Packer can be used with both JSON and HCL2 templates. If you are using JSON templates we recommend getting started with 
> the [HCP Packer environment variables](#hcp-packer-environment-variables) and then, when ready, migrate from JSON to HCL2 templates. 

Reviewer: ```suggestion
You can use HCP Packer with both JSON and HCL2 templates. If you are using JSON templates we recommend getting started with 
the [HCP Packer environment variables](#hcp-packer-environment-variables) and then migrating to HCL when possible. 
```

> ~> Refer to the [HCP Packer Template Configuration](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration) documentation for details 

Reviewer: ```suggestion
This page summarizes the methods you can use to connect JSON and HCL2 templates to the HCP Packer registry. It also provides a full list of HCP Packer environment variables. Refer to the [HCP Packer Template Configuration](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration) page in the HCP Packer documentation for full configuration details and examples.
```

Author: Very nice. I like this approach.

> on using HCL2 templates to push image metadata to an active HCP Packer registry.

Reviewer: ```suggestion
```

## HCP Packer Environment Variables

> The following environment variables allow you to configure Packer to push image metadata to an active registry without changing your template. 
> 
> Environment variables are the only way to configure HCP Packer from a JSON template. 
> 
> You must set following environment variables to enable Packer to push metadata to a registry.

Reviewer: ```suggestion
The following environment variables let you configure Packer to push image metadata to an active registry without changing your template. You can use environment variables with both JSON and HCL2 templates. Refer to [Basic Configuration With Environment Variables](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration#basic-configuration-with-environment-variables) in the HCP Packer documentation for complete instructions and examples. 

You must set following environment variables to enable Packer to push metadata to a registry.
```

Author: I like the idea of having the Basic Configuration section.

- `HCP_CLIENT_ID`         - The HCP client ID of the HashiCorp Cloud Platform service principle that will be authenticating Packer to the HCP Packer registry.

- `HCP_CLIENT_SECRET`    - The HCP client secret of the HashiCorp Cloud Platform service principle that will be authenticating Packer to the HCP Packer registry.

- `HCP_PACKER_BUCKET_NAME` -  The name of the Image bucket where Packer will store information about each artifact from a build within the configured template. 

You can set these additional environment variables to control how metadata is pushed to the registry.

- `HCP_PACKER_BUILD_FINGERPRINT` - A unique identifier assigned to a build run to determine if metadata for a template build on the registry is complete. By default, the HEAD Git SHA 
for the template file is used to create a fingerprint. If the template is not version controlled, this environment variable must be set to a unique value before executing a build on the template. 

- `HCP_PACKER_REGISTRY`    -  Allows users to toggle HCP Packer publishing off. Allowed values are [0|OFF].

## HCP Packer Registry Block

> For HCL2 templates, the recommended approach for configuring HCP Packer is within the template via the `hcp_packer_registry` block,
> as it lets you customize the metadata Packer sent to HCP Packer registry. 

Reviewer: ```suggestion
The only metadata that Packer can infer from a template with the basic configuration are the build name and build fingerprint. For HCL2 templates, we recommend adding the `hcp_packer_registry` block to your template so that you can customize the metadata that Packer sends to the registry. 

The `hcp_packer_registry` block is only available for HCL2 Packer templates. There is no [ `PACKER_CONFIG`](docs/configure#packer-s-config-file) equivalent for JSON.
```

Author: Should we maybe add "environment variables" to the sentence to be more explicit on which basic configuration we are referring to? 

```
The only metadata that Packer can infer from a template with the basic environment variables configuration are the build name and build fingerprint.
```

> ~> Refer to [`hcp_packer_registry`](docs/templates/hcl_templates/blocks/build/hcp_packer_registry) for a full list configuration arguments.
> 
> The `hcp_packer_registry` block is not supported by JSON templates; nor is there a [ `PACKER_CONFIG` ](docs/configure#packer-s-config-file) equivalent. 
> 
> Usage of the `hcp_packer_registry` block requires that the environment variables `HPC_CLIENT_ID` and `HCP_CLIENT_SECRET` be set. The existence of any other variable such
> as `HCP_PACKER_BUCKET_NAME` or `HCP_PACKER_BUILD_FINGERPRINT` will override any configuration settings defined within the template. 

Reviewer: ```suggestion
Refer to [`hcp_packer_registry`](docs/templates/hcl_templates/blocks/build/hcp_packer_registry) for a full list of configuration arguments. Refer to [Custom Configuration](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration#custom-configuration) in the HCP Packer documentation for information and examples about how to customize image metadata.
```

Author: Okay so the goal here is just to redirect them to the page with all the info.

---
description: |

>   Packer can be configured to publish artifact metadata for all completed builds to an active HCP Packer registry.
>   The configuration for HCP Packer is managed by environment variables for legacy JSON templates or via a hcp_packer_registry
>   block for HCL2 templates. 

Reviewer: ```suggestion
  You can configure Packer to publish metadata for all completed builds to an active HCP Packer registry. Legacy JSON templates can connect to the registry using environment variables. HCL2 templates can connect using an hcp_packer_registry block. 
```

> page_title: Configuring Packer for HCP Packer

Reviewer: I think we could just go with "Configuring HCP Packer " - I feel like the extra Packer is a bit confusing

Author: 💯 thanks for calling that out.

> page_title: Configuring Packer for HCP Packer

Author: ```suggestion
page_title: Configuring HCP Packer
```

---

# Configuring Packer for HCP Packer

> HCP Packer stores metadata about your Packer images so that you can track updates, use the most up-to-date base images, and deploy the most up-to-date downstream images. 
> For more details on How it Works see the [HashiCorp Cloud Platform Packer documentation](https://cloud.hashicorp.com/docs/packer#what-is-hcp-packer) 
> 
> You can configure any Packer template to push metadata to the HCP Packer registry, but we recommend that you use HCL2 Packer templates. 
> The HCL2 support in Packer allows you to codify how Packer interacts with the registry and customize your image metadata. 
> 
> Users wishing to use HCP Packer with JSON are encouraged to get started with the environment variables and eventually work to migrate from JSON to HCL2 configuration
> templates. 

Reviewer: ```suggestion
[HCP Packer](https://cloud.hashicorp.com/docs/packer) stores metadata about your Packer images so that you can track updates, use the most up-to-date base images, and deploy the most up-to-date downstream images. 

You can configure any Packer template to push metadata to the HCP Packer registry, but we recommend that you use HCL2 Packer templates. The HCL2 support in Packer allows you to codify how Packer interacts with the registry and customize your image metadata. Refer to [Template Configuration](https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration) in the HCP Packer documentation for details. 

If you want to use HCP Packer with JSON, we recommend first connecting with [HCP Packer environment variables](#hcp-packer-environment-variables) and then migrating from JSON to HCL2 configuration templates. 
```

## HCP Packer Environment Variables

HCP Packer configuration through environment variables is currently the only way to publish artifacts from a Packer JSON template. 

The HCP Packer environment variables are made available for users who wish to publish their artifacts to an active HCP Packer registry
without making any changes to their existing template. Making it an ideal solution for users looking to get started with an existing template; both JSON and HCL2.

> HCP Packer environment variables must be configured before executing a Packer build command. They work with all template types but are required for legacy JSON templates. 

Author: I really like the wording here. Thank you!

> ** Required Environment Variables**

Reviewer: One thought on this is --> we have a really clear list of requirements in the HCP Packer docs. Should we replicate those here? For example, we talk about the specific version of Packer you need, etc. I worry that if we don't, users may try to configure their connection from this alone and possibly get stuck. 

This is partially why (refer to my comment on the PR) I think it could be more beneficial to fold this into the existing HCP Packer docs. Regardless, I think we need to let users know about all the requirements!
```suggestion
You must set the following environment variables to enable Packer to push metadata to the HCP Packer registry.
```

- `HCP_CLIENT_ID`         - The HCP client ID of the HashiCorp Cloud Platform service principle that will be authenticating Packer to the HCP Packer registry.

- `HCP_CLIENT_SECRET`    - The HCP client secret of the HashiCorp Cloud Platform service principle that will be authenticating Packer to the HCP Packer registry.

- `HCP_PACKER_BUCKET_NAME` -  The name of the Image bucket where Packer will store information about each artifact from a build within the configured template. 

** Optional Environment Variables**

- `HCP_PACKER_BUILD_FINGERPRINT` -  A unique identifier assigned to a build run to determine if metadata for a template build on the registry is complete. A fingerprint can 
not be more than 32 characters long.

- `HCP_PACKER_REGISTRY`    -  Allows users to toggle HCP Packer publishing off. Only valid when all required variables have been set. Allowed values are [0|OFF].

> ~> **Build Fingerprint**: By default, Packer uses the HEAD Git SHA for the template file to create a fingerprint. 

Reviewer: can we move this up to be under the build fingerprint? OR maybe we can incorporate this information into the description? I just want it to feel cohesive :)

Author: Good call ✅

If your template is not version controlled, you must set the HCP_PACKER_BUILD_FINGERPRINT environment variable and set it to a unique value each time you call packer build. 
Packer will not build the template unless it detects a build fingerprint.

## HCP Packer Registry HCL Block

> HCL2 templates, along with support for environment variables, can be configured using a `hcp_packer_registry` block, lets you customize the metadata Packer sends to HCP Packer Registry. It configures the details of an image that is created or updated within the HCP Packer registry. 

Reviewer: ```suggestion
For HCL2 templates, you can also configure HCP Packer using an `hcp_packer_registry` block. We recommend this approach because it lets you customize the metadata Packer sends to HCP Packer Registry. 
```

> The HCP Packer registry block in HCL is the preferred method for configuring Packer for the HCP Packer registry,

Reviewer: We've got a lot of detailed information about this and the bucket labels, etc. in the HCP docs that I don't think we should replicate here (but that users may want to know!). I wonder if we can shorten this and just link out to the existing content we have in the HCP docs?

Reviewer: Or, maybe just link more quickly to the registry blockpage --> it looks like it has an expanded version of this same content. The thing I want to avoid is giving users JUST enough that they try to do things with this information, but not enough to give them all the context they need to set up a good workflow/ to avoid stubbing their toe.

Author: This is a fair concern and one I struggled with. I like the idea of linking sooner to not duplicate and guiding folks to the information they need to build a solid workflow. I would venture to say that the `hcp_packer_block` page should also link out to the complete documentation at https://cloud.hashicorp.com/docs/packer/store-image-metadata/template-configuration. This way we can really hone in on the clear list of requirements in the HCP Packer docs. 

We can do the same for the JSON docs once the HCP Packer docs have been updated with the JSON details.

as it allows for user customizable metadata. This block can not be used with legacy JSON template; nor is there a PACKER_CONFIG equivalent. 

Usage of the `hcp_packer_registry` block requires that the environment variables `HPC_CLIENT_ID` and `HCP_CLIENT_SECRET` be set. The existence of any other variable such
as `HCP_PACKER_BUCKET_NAME` or `HCP_PACKER_BUILD_FINGERPRINT` will override any configuration settings defined within the `hcp_packer_registry` block. 

Below is an example of the `hcp_packer_registry` block. 
```hcl
source "happycloud" "macos" {
  os = "macos_amd64"
}

build {
   hcp_packer_registry {
    bucket_name = "ios-dev"

    description = " A description about the image which artifact is being published to HCP Packer Registry."

    bucket_labels = {
      "team" = "ios-development",
      "os"   = "macos"
    }

    build_labels = {
      "xcode"   = "11.3.0"
      "version" = "Big Sur"
    }
  }

  sources = ["source.happycloud.macos"]
}

```

> See [`hcp_packer_registry`](docs//templates/hcl_templates/blocks/build/hcp_packer_registry) for a full list configuration arguments.

Reviewer: We always want to say "refer to" to avoid ableist language :) 
```suggestion
Refer to [`hcp_packer_registry`](docs//templates/hcl_templates/blocks/build/hcp_packer_registry) for a full list configuration arguments.
```

Author: Got it. Thank you for calling this out.

- `HCP_CLIENT_SECRET`    - The HCP client secret of the HashiCorp Cloud Platform service principle that will be authenticating Packer to the HCP Packer registry.

- `HCP_PACKER_BUCKET_NAME` -  The name of the Image bucket where Packer will store information about each artifact from a build within the configured template. 

** Optional Environment Variables**

- `HCP_PACKER_BUILD_FINGERPRINT` -  A unique identifier assigned to a build run to determine if metadata for a template build on the registry is complete. A fingerprint can 
not be more than 32 characters long.

- `HCP_PACKER_REGISTRY`    -  Allows users to toggle HCP Packer publishing off. Only valid when all required variables have been set. Allowed values are [0|OFF].

~> **Build Fingerprint**: By default, Packer uses the HEAD Git SHA for the template file to create a fingerprint. 
If your template is not version controlled, you must set the HCP_PACKER_BUILD_FINGERPRINT environment variable and set it to a unique value each time you call packer build. 
Packer will not build the template unless it detects a build fingerprint.

## HCP Packer Registry HCL Block

> HCL2 templates, along with support for environment variables, can be configured using a `hcp_packer_registry` block, lets you customize the metadata Packer sends to HCP Packer Registry.
> It configures the details of an image that is created or updated within the HCP Packer registry. The HCP Packer registry is the preferred method for configuring Packer for the HCP Packer registry,
> as it allows for user customizable metadata. This block can not be used with legacy JSON template; nor is there a PACKER_CONFIG equivalent. 

Author: ```suggestion
HCL2 templates, along with support for environment variables, can be configured using a `hcp_packer_registry` block, lets you customize the metadata Packer sends to HCP Packer Registry. It configures the details of an image that is created or updated within the HCP Packer registry. 

The HCP Packer registry block in HCL is the preferred method for configuring Packer for the HCP Packer registry,
as it allows for user customizable metadata. This block can not be used with legacy JSON template; nor is there a PACKER_CONFIG equivalent. 
```

Usage of the `hcp_packer_registry` block requires that the environment variables `HPC_CLIENT_ID` and `HCP_CLIENT_SECRET` be set. The existence of any other variable such
as `HCP_PACKER_BUCKET_NAME` or `HCP_PACKER_BUILD_FINGERPRINT` will override any configuration settings defined within the `hcp_packer_registry` block. 

Below is an example of the `hcp_packer_registry` block. 
```hcl
source "happycloud" "macos" {
  os = "macos_amd64"
}

build {
   hcp_packer_registry {
    bucket_name = "ios-dev"

    description = " A description about the image which artifact is being published to HCP Packer Registry."

    bucket_labels = {
      "team" = "ios-development",
      "os"   = "macos"
    }

    build_labels = {
      "xcode"   = "11.3.0"
      "version" = "Big Sur"
    }
  }

  sources = ["source.happycloud.macos"]
}

```

See [`hcp_packer_registry`](docs//templates/hcl_templates/blocks/build/hcp_packer_registry) for a full list configuration arguments.

> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Author: ```suggestion
```
